compiler/rustc_const_eval/src/interpret/operator.rs

```diff
@@ -127,8 +127,20 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
         // Shift ops can have an RHS with a different numeric type.
         if bin_op == Shl || bin_op == Shr {
-            let signed = left_layout.abi.is_signed();
             let size = u128::from(left_layout.size.bits());
+            // Even if `r` is signed, we treat it as if it was unsigned (i.e., we use its
+            // zero-extended form). This matches the codegen backend:
+            // <https://github.com/rust-lang/rust/blob/f047af24b3b90f19b89fac80822acd69613b89ec/compiler/rustc_codegen_ssa/src/base.rs#L315-L317>.
+            // The overflow check is also ignorant to the sign:
+            // <https://github.com/rust-lang/rust/blob/8fae33d9b21e1e29da65506a2d6ac5bbbb3c4a86/compiler/rustc_codegen_ssa/src/mir/rvalue.rs#L728>.
```

Comment:
I don't like permalinks too much in comments, especially two of them so close to each other but at different commits.

I don't have a good solution that won't cause the information to go stale though. At best we could have some soft of rustdoc link that will get complained about if it breaks and then link to the function in codegen.

Oh well, let's merge it, but please use the same commit id.